platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/ConventionMappingFileSystemLocationPropertyProxy.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.provider;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.provider.Provider;
+
+/**
+ * This class is here just to support FileSystemLocation properties in the convention mapping.
+ *
+ * Should be removed once ConventionMapping is removed.
+ */
+@NonNullApi
+public interface ConventionMappingFileSystemLocationPropertyProxy {
+    void conventionFromAny(Provider<Object> file);
+}
```

Comment:
:x: I don't think we need this type. We can simply do what we need in `ConventionAwareHelper.mapConventionOn()` when we detect a `FileSystemLocationProperty`, no?

Reply:
I added this type because we cannot create a Directory or a RegularFile inside `ConventionAwareHelper`, we can't access any factory there, or at least am not aware of a way to do that.

That is why we then pass `File` to a DirectoryProperty/RegularFileProperty via this method and there we can then do a conversion from a `File` to a `Directory` or a `RegularFile`

Reply:
From what I noticed around, whenever we need implementations of an API type to support some internal protocol that we don't want to expose with API users, we create a `SomeApiTypeInternal` interface that (optionally) extends `SomeApiType` and is implemented by all implementations of `SomeApiType`. 

For instance, `ProjectInternal`, `PropertyInternal`, `ProviderInternal`. 

Would it make sense to follow that here, @asodja?

Reply:
Good point. I introduced `FileSystemLocationPropertyInternal extends FileSystemLocationProperty`. This also makes more sense now and it's more consisten with our code style compared to previous `ConventionMappingFileSystemLocationPropertyProxy`.